### DISCUSSION TITLE
Smooth scrolling

### DISCUSSION THREAD

**Discussion by JoseConseco** [helix-editor/helix#10220] [Ideas]
I did not saw anyone requesting this.

https://user-images.githubusercontent.com/13521338/154261469-f8190bbe-4b80-417f-b9cd-5a041ab4c968.mp4

Document navigation without smooth scroll is very confusing IMO.  Would love to se this in helix

**Comment by kirawi** [comment]
Does changing the value of `scroll-lines` from https://docs.helix-editor.com/configuration.html#editor-section help?

**Comment by JoseConseco** [comment]
Actually mouse scrolling is ok, but what I wished for was smooth scroll for:  half page up/down,  full page up/down hotkeys.
For nvim I use: [neoscroll](https://github.com/karb94/neoscroll.nvim)
Navigation without it, is very confusing IMO as it is impossible to track moved code with eyes. If you try it, then u wont be able to go back. IMO, lack of smooth page navigation, is reason why people do not use  (half) page up or down that much.

**Comment by archseer** [comment]
> is very confusing IMO

Personally I find smooth scrolling disorienting. It's also very inefficient because it will trigger lots of full screen redraws.

To get this into `helix-term` it would require a bunch of internal changes so we could animate the transition between positions. Right now we assume that an action gets a single redraw.

We will support smooth scrolling on alternate renderers one day though.

**Comment by txtyash** [comment]
Made a poll: https://github.com/helix-editor/helix/discussions/3746

**Comment by SteveLauC** [comment]
In vim or neovim, one can use `CTRL-e` and `CTRL-y` to do this. 

In helix, my current workaround is:
1. To smoothly scroll down:
    1. use `zb` to move the cursor to the bottom
    2. then use `j` to scroll down.
3. To smoothly scroll up:
    1. use `zt` to move the cursor to the top
    2. then use `k` to scroll up.

Or you can add a key mapping for this combination:
```toml
[keys.normal]
C-e = ["align_view_bottom", "move_line_down" ]
C-y = ["align_view_top", "move_line_up" ]
```

-----
Really love to see this feature added in helix:)

**Comment by batalooppak** [comment]
> In vim or neovim, one can use `CTRL-e` and `CTRL-y` to do this.
> 
> In helix, my current workaround is:
> 
> 1. To smoothly scroll down:
>    
>    1. use `zb` to move the cursor to the bottom
>    2. then use `j` to scroll down.
> 2. To smoothly scroll up:
>    
>    1. use `zt` to move the cursor to the top
>    2. then use `k` to scroll up.
> 
> Or you can add a key mapping for this combination:
> 
> ```toml
> [keys.normal]
> C-e = ["align_view_bottom", "move_line_down" ]
> C-y = ["align_view_top", "move_line_up" ]
> ```
> 
> Really love to see this feature added in helix:)

Thank you Steve. This works better to me. test both version to your liking:
```toml
[keys.normal]
# C-e = "scroll_down"
# C-y = "scroll_up"
C-e = ["scroll_down", "move_line_down"]
C-y = ["scroll_up", "move_line_up"]
```
**NOTE**: When cursor is in the first line it jumps to the 8th line. It doesn't happen from bottom to top

**ADDENDUM**: The cursor gets to the 8th line, because it's taking into account the scrolloff (default:5) + the movie_line_down. In step-by-step is like this: We are in line 1, we scroll_down, which leaves us at line 2, we move the cursor one more time with move_line_down leaves us at line 3 and the scrolloff (default: 5) leaves us at line 8.

https://github.com/helix-editor/helix/blob/2b7e7c80eb4bdc0c7283f94efd6079992840753b/helix-term/src/commands.rs#L1555

and that's why it doesn't happen at the bottom when scrolling up, because the buffer is allowed to be scrolled past the EOF.

**ADDENDUM 2**: Just learn about 'sticky view mode'. With sticky mode you just do Shift-z and then you can scroll with j/k keys.

**Comment by thomasschafer** [comment]
You can enable smooth scrolling with this plugin: https://github.com/thomasschafer/smooth-scroll.hx
